uaclient/version.py

```diff
@@ -29,8 +39,43 @@ def get_version():
     if os.path.exists(os.path.join(topdir, ".git")):
         cmd = ["git", "describe", "--abbrev=8", "--match=[0-9]*", "--long"]
         try:
-            out, _ = system.subp(cmd)
+            out, _ = subp(cmd)
             return out.strip()
-        except exceptions.ProcessExecutionError:
+        except ProcessExecutionError:
             pass
     return __VERSION__
+
+
+def get_last_known_candidate():
+    if (
+        not os.path.exists(CANDIDATE_CACHE_PATH)
+        or os.stat(CANDIDATE_CACHE_PATH).st_mtime < get_apt_cache_time()
+    ):
+        try:
+            policy = run_apt_cache_policy_command(
+                package="ubuntu-advantage-tools"
+            )
+            match = re.search(CANDIDATE_REGEX, policy)
+            if match:
+                candidate_version = match.group("candidate")
+                os.makedirs(UAC_TMP_PATH, exist_ok=True)
+                with open(CANDIDATE_CACHE_PATH, "w") as f:
+                    f.write(candidate_version)
+                os.chmod(CANDIDATE_CACHE_PATH, 0o777)
+        except Exception:
+            pass
+
+    try:
+        with open(CANDIDATE_CACHE_PATH, "r") as f:
+            return f.read()
+    except Exception:
+        pass
+
+    return ""
```

Comment:
IMO we should either return `Optional[str]` (so `None` here), or document that it returns `""` if there is no candidate